File: operating_systems__COSC310/lab5/lab5.c
/*
 * TJ Liggett
 * 3_10_2019
 * COSC 310
 * Lab 5
 * main.c
 */


#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define MAX_STRING_LEN 80

void help();
void dump(int fltCnt, int v[], int p[]);
void textbook(int verb, int *fltCnt, int v[], int p[]);
int mmu(int *fltCnt, int loc, int v[], int p[], int verb);
void pageFault(int pageNum, int v[], int p[], int verb);


void main()
{

    char command[MAX_STRING_LEN];   /* User command string */
    int loc;                         /* User command int space */
    int virtAdd[16];                /* Virtual address page table */
    int physAdd[8];                 /* Physical address page table */
    int verb = 0;           /* boolean value set to status of verbose mode (0 is false) */
    int i;                          /* Loop counter */
    int fltCnt = 0;                      /* Count of faults in program */

    // Assign mapping assignments to virtual pages 0-7
    // set physical page reference counts to 0
    for(i = 0; i < 8; ++i)
    {
        physAdd[i] = 0;
        virtAdd[i] = i;
    }
    //Declare virtual pages 8-15 as unmapped.
    for(i = 8; i < 16; ++i) virtAdd[i] = 8;

    printf("COSC 310 Lab 5 by TJ Liggett. For help input ?\n");

    //Handle user commands
    while(free)
    {
        //Prompt user for command
        printf("tjlab5> ");
        scanf("%s", command);

        //Determine operation based on command
        if(strcmp(command, "q") == 0) break;
        else if(strcmp(command, "?") == 0) help();

        //Set verbose mode on/off
        else if(strcmp(command, "on") == 0)
        {
            verb = 1;
            printf("   Verbose Mode ON\n");
        }
        else if(strcmp(command, "off") == 0)
        {
            verb = 0;
            printf("   Verbose Mode OFF\n");
        }
        //dump current page tables
        else if(strcmp(command, "dump") == 0) dump(fltCnt, virtAdd, physAdd);
        //set up tables like Figure 3-9 in the textbook
        else if(strcmp(command, "textbook") == 0) textbook(verb, &fltCnt, virtAdd, physAdd);
        //Interpret address requests/invalid commands
        else
        {

            if(strcmp(command, "0") == 0) /* Ensures address is not zero */
            {
                loc = 0;
                printf("   %d --> %d\n", loc, mmu(&fltCnt, loc, virtAdd, physAdd, verb));
            }
            else
            {
                loc = atoi(command); /* This will return integer if command is integer, 0 else */
                if(loc < 1 || loc > 65535) printf("Invalid Command. Enter ? for help.\n");
                else printf("   %d --> %d\n", loc, mmu(&fltCnt, loc, virtAdd, physAdd, verb));
            }

        }


    }

}


//Outputs a help menu on console
void help()
{
    printf("Commands:\n");
    printf("   q\t\tQuit the program\n");
    printf("   ?\t\tPrint this help screen\n");
    printf("   on\t\tTurns Verbose Mode On\n");
    printf("   off\t\tTurns Verbose Mode Off\n");
    printf("   dump\t\tDumps the current Page Table\n");
    printf("   textbook\tSets up pages as in Fig 3-9 in text\n");
    printf("   1729\t\t(Or any number -- decode this address)\n");
}

//Outputs page tables on console
void dump(int fltCnt,   /* Count of page faults */
        int v[],        /* virtual address map array*/
        int p[])        /* physical address reference array*/
{
    int i = 0;          /* Loop counter */


    printf("     Virtual-Address   Physical-Address\n");
    printf("\tIndex  MapsTo\tIndex  RefCnt\n");
    for(i = 15; i > 7; --i) printf("\t%4d%4d\n", i, v[i]);
    for(i = 7; i >= 0; --i) printf("\t%4d%4d\t%4d%4d\n", i, v[i], i, p[i]);
    printf("     Number of Page-Faults: %d\n", fltCnt);
}

//Sets up pages as in Fig 3-9 in text
//NOTE: textbook resets the page-fault count, as it is assumed there are 0 faults
//      at the beginning of the textbook example
void textbook(int verb,    /* Verbose mode boolean */
              int *fltCnt, /* Fault count */
              int v[],     /* virtual address map array*/
              int p[])     /* physical address reference array*/
{
    *fltCnt = 0;
    v[0] = 2;
    v[1] = 1;
    v[2] = 6;
    v[3] = 0;
    v[4] = 4;
    v[5] = 3;
    v[6] = 8;
    v[7] = 8;
    v[8] = 8;
    v[9] = 5;
    v[10] = 8;
    v[11] = 7;
    v[12] = 8;
    v[13] = 8;
    v[14] = 8;
    v[15] = 8;
    for(int i = 0; i<8; ++i) p[i] = 0;
    if(verb == 1) printf("Set up pages as in Fig 3-9 in text\nNOTE: Page-Fault count was reset\n");
}

//Fetch physical memory location from virtual location
int mmu(int *fltCnt,     /* Page-Fault count */
        int loc,        /* Virtual address */
        int v[],        /* Virtual address page table */
        int p[],        /* Physical address page table */
        int verb)       /* Verbose mode boolean (0 is false)*/
{
    int offset;         /* Page offset for address */
    int pageTemp;       /* Temporary page number */
    int vPage, pPage;   /* Virtual and Physical page number */
    int pAdd;           /* Physical address to return */


    //Virtual page and offset of address are found
    offset = loc & 0x0FFF;
    pageTemp = loc & 0xF000;
    vPage = pageTemp >> 12;
    if(verb == 1) printf("Address:  %d offset:  %d blkNum:  %d PageTableLoc:  %d\n", loc, offset, vPage, v[vPage]);

    //Checks for page-fault. If yes, then execute page-fault operation and increment fault counter
    if(v[vPage] == 8)
    {
        pageFault(vPage, v, p, verb);
        *fltCnt = (*fltCnt + 1);
        if(verb == 1) printf("After-Trap -- Address:  %d offset:  %d blkNum:  %d PageTableLoc: %d\n", loc, offset, vPage, v[vPage]);
    }
    //Finds physical address and returns
    pPage = v[vPage];
    pAdd = (pPage << 12) | offset;
    p[pPage]++;
    return pAdd;
}

//Handles page faults in memory
void pageFault( int pageNum,/* Virtual page number*/
                int v[],    /* Virtual address page table */
                int p[],    /* Physical address page table */
                int verb)   /* Verbose mode boolean (0 is false) */
{
    int low = 0;            /* The page number of page with lowest reference count */


    //Finds page with lowest reference count and evicts from the table
    for(int i = 1; i < 8; ++i) if(p[i] < p[low]) low = i;
    if(verb == 1) printf("PageFrame %d had reference count of %d. Evicted from Page Table\n", low, p[low]);

    //Unmaps virtual page with lowest reference count (least used)
    for(int i = 0; i < 15; ++i)
        if(v[i] == low)
        {
            v[i] = 8;
            if(verb == 1) printf("Virtual page table at location %d set to 8 (unmapped)\n", i);
        }

    //Map virtual page needed to least used physical page
    v[pageNum] = low;
    p[low] = 0;
}

File: operating_systems__COSC310/lab5/lab5.java
/*
 * TJ Liggett
 * 3_7_2019
 * COSC 310
 * Lab 5
 * lab5.java
 */

import java.util.*;
import java.io.*;

public class lab5
{

    public static boolean verb;
    public static int[] virtAdd = new int[16];
    public static int[] physAdd = new int[8];

    public static void main(String[] args)
    {
        Scanner input = new Scanner(System.in);
        String command = "";
        verb = true;

        for(int i = 0; i<8; ++i) virtAdd[i] = i;
        for(int i = 8; i<16; ++i) virtAdd[i] = 8;
        for(int i = 0; i<8; ++i) physAdd[i] = 0;



        System.out.println("COSC 310 Lab 5 by TJ Liggett. Enter ? for help.");
        while(true)
        {
            System.out.print("lab5> ");
            command = input.nextLine();
            command = command.trim();
            if(command.equals("q")) break;
            if (command.matches("\\d+"))
            {
                int vAdd = Integer.parseInt(command);
                if(vAdd >= 65536) System.out.printf("Address is out of bounds!! \n");
                else System.out.printf("   %d --> %d\n", vAdd, mmu(vAdd));
            }
            else switch(command)
            {
                case "?": help();
                break;
                case "on": verbose(true);
                break;
                case "off": verbose(false);
                break;
                case "dump": dump();
                break;
                case "textbook": textbook();
                break;
                default: System.out.printf("%s not a valid command. Press ? for help.\n", command);
                break;

            }

        }
    }

    public static void help()
    {
        System.out.printf("Commands:\n");
        System.out.printf("   q\t\tQuit the program\n");
        System.out.printf("   ?\t\tPrint this help screen\n");
        System.out.printf("   on\t\tTurns Verbose Mode On\n");
        System.out.printf("   off\t\tTurns Verbose Mode Off\n");
        System.out.printf("   dump\t\tDumps the current Page Table\n");
        System.out.printf("   textbook\tSets up pages as in Fig 3-9 in text\n");
        System.out.printf("   1729\t\t(Or any number -- decode this address)\n");
    }

    public static void verbose(boolean v)
    {
        if(v)
        {
            verb = true;
            System.out.println("   Verbose mode ON");
        }else{
            verb = false;
            System.out.println("   Verbose mode OFF");
        }
    }

    public static void dump()
    {
        System.out.println("     Virtual-Address   Physical-Address");
        System.out.println("\tIndex  MapsTo\tIndex  RefCnt");
        for(int i = 15; i > 7; --i) System.out.printf("\t%4d%4d\n", i, virtAdd[i]);
        for(int i = 7; i >= 0; --i) System.out.printf("\t%4d%4d\t%4d%4d\n", i, virtAdd[i], i, physAdd[i]);
    }

    //Set up virtual mapping as in Fig 3-9 in text and reset ref Counts.
    public static void textbook()
    {
        virtAdd[0] = 2;
        virtAdd[1] = 1;
        virtAdd[2] = 6;
        virtAdd[3] = 0;
        virtAdd[4] = 4;
        virtAdd[5] = 3;
        virtAdd[6] = 8;
        virtAdd[7] = 8;
        virtAdd[8] = 8;
        virtAdd[9] = 5;
        virtAdd[10] = 8;
        virtAdd[11] = 7;
        virtAdd[12] = 8;
        virtAdd[13] = 8;
        virtAdd[14] = 8;
        virtAdd[15] = 8;
        for(int i = 0; i<8; ++i) physAdd[i] = 0;
        if(verb) System.out.println("Set up pages as in Fig 3-9 in text");
    }

    //
    public static int mmu(int vAdd)
    {
        int offset = vAdd%4096;
        int blkNum = vAdd/4096;
        if(verb) System.out.printf("Address:  %d offset:  %d blkNum:  %d PageTableLoc:  %d\n", vAdd, offset, blkNum, virtAdd[blkNum]);
        if(virtAdd[blkNum] == 8)
        {
            pageFault(blkNum);
            if(verb) System.out.printf("After-Trap -- Address:  %d offset:  %d blkNum:  %d PageTableLoc: %d\n", vAdd, offset, blkNum, virtAdd[blkNum]);
        }
        int physPage = virtAdd[blkNum];
        int pAdd = (physPage * 4096) + offset;
        physAdd[physPage]++;
        return pAdd;
    }

    public static void pageFault(int blkNum)
    {
        int low = 0;
        for(int i = 1; i < 8; ++i) if(physAdd[i] < physAdd[low]) low = i;
        if(verb) System.out.printf("PageFrame %d had reference count of %d. Evicted from Page Table\n", low, physAdd[low]);
        int unmap = 0;
        for(int i = 0; i < 15; ++i)
            if(virtAdd[i] == low)
            {
                virtAdd[i] = 8;
                if(verb) System.out.printf("Virtual page table at location %d set to 8 (unmapped)\n", i);
            }
        virtAdd[blkNum] = low;
        physAdd[low] = 0;
    }


}